Add functions, roles, and object models to draft

mickenordin opened the pull request:
This updates IETF-RFC.md with a significantly expanded framework for describing OCM using ASCII diagrams in the vain of https://datatracker.ietf.org/doc/rfc3290/ as referenced by https://datatracker.ietf.org/doc/rfc3444/.

It introduces formal definitions of OCM functions, clarifies the roles of Sending and Receiving Servers, and adds detailed object models for Address Books, Contacts, Invites, Providers, Shares, Users, and Resources. The change reorganizes and expands terminology, adds diagrams to illustrate relationships, and alphabetizes and harmonizes defined terms. Legacy duplicated sections are removed or merged into the new structure.

mickenordin committed:
Add functions, roles, and object models to draft

This updates IETF-RFC.md with a significantly expanded conceptual
framework for OCM. It introduces formal definitions of OCM functions,
clarifies the roles of Sending and Receiving Servers, and adds detailed
object models for Address Books, Contacts, Invites, Providers, Shares,
Users, and Resources. The change reorganizes and expands terminology,
adds diagrams to illustrate relationships, and alphabetizes and
harmonizes defined terms. Legacy duplicated sections are removed or
merged into the new structure.

Signed-off-by: Micke Nordin <kano@sunet.se>

KrausMatthias commented on IETF-RFC.md:
```diff
@@ -85,60 +85,443 @@ NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
+
+A Sending Server is an OCM Provider that holds Resources and exposes
+APIs to allow access to them.  It allows its users to create _Shares_
+or _Invites_ to give other users access to those Resources. A Sending
```
or _Invites_ appears to be duplicated here?

mickenordin replied:
I don't see it?

mickenordin replied:
do you mean it should be invites, below, without italics?

KrausMatthias replied:
Maybe I'm overly picky. In the sentence 

> It allows its users to create _Shares_ **or _Invites_** to give other users access to those Resources.

Invites don't have anything to do with Resource Access. Also creating invites is not optional there, but is in the next sentence.

KrausMatthias commented on IETF-RFC.md:
```diff
@@ -85,60 +85,443 @@ NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
+An _OCM Provider_ MAY offer its _Users_ an address book tool, where OCM
+Addresses can be stored over time in a labeled and/or searchable way.
+This decouples the act by which the OCM Address string is passed into
+the Sending Server's database from the selection of the Receiving Party
```
Should Receiving Party be _Receiving Party_ ?

KrausMatthias commented on IETF-RFC.md:
```diff
@@ -85,60 +85,443 @@ NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
+* __trusted__: Indicates if contact has been added through the
+               Invite flow (trusted), via direct entry (untrusted)
+               or is blocked
+* __userID__: The identifier of the contact at their OCM Server
```
userID is missing in the plot. Why would you need this in addition to the ocmAddress?

mickenordin replied:
I was think about this question when writing it, and struggled abit, since the same model is used for both local and remote users. But maybe it makes sense to drop the local uid and just use the ocm address.

MahdiBaghbani replied:
I'm also in favor of using the OCM Adress in here

mickenordin replied:
All our api's use userID and/or (userID and provider):

* https://github.com/cs3org/OCM-API/blob/7821ae8dc982fe9e752cd0c84f75dc1015460649/spec.yaml#L815

* https://github.com/cs3org/OCM-API/blob/7821ae8dc982fe9e752cd0c84f75dc1015460649/spec.yaml#L785

So maybe we ought to do that here as well?

mickenordin replied:
I implemented invites today, in my toy app, and there is alot of this: https://code.smolnet.org/micke/amity/src/commit/2c26625ee0d45ab42bcaf249fa9314eab08d2228/app/paths/invite_accepted.py#L32

https://code.smolnet.org/micke/amity/src/commit/2c26625ee0d45ab42bcaf249fa9314eab08d2228/app/paths/accept_invite.py#L96

So I think I am in favor of using userID instead of the ocm-address

glpatcern replied:
Indeed, we also store in our data model an `opaque_user_id` and an `idp` as opposed to the OCM Address, I think it makes sense from a modelling perspective.

KrausMatthias commented on IETF-RFC.md:
```diff
@@ -85,60 +85,443 @@ NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
+
+* __Blocked__: Contact blocked from sending Shares
+* __Trusted__: Established through completed Invite flow
+* __Untrusted__: Added directly without Invite verification
```
Not totally convinced of this. What if I want to control who a user may share with? E.g. to avoid accidental data leaks?  Just a bit worried this adds to much detail which is not directly relevant to interoperability.

mickenordin replied:
I could remove the states and just mention in text that an implementation MAY treat invites differently from direct entry.

KrausMatthias commented on IETF-RFC.md:
```diff
@@ -85,60 +85,443 @@ NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
+* __Accepted__: Invite successfully accepted by Invite Receiver
+* __Expired__: Invite no longer valid after a certain time
+* __Pending__: Invite created but not yet accepted
+* __Revoked__: Invite invalidated by Invite Sender before acceptance
```
I don't know if I've missed something here as I haven't implemented Invites so far, but why would I store "Non-Pending" invites? Acceptance leads to a trusted entry in the address book. What would I use expired or revoked tokens for?

mickenordin replied:
Same here, maybe just remove the states then?

KrausMatthias commented on IETF-RFC.md:
```diff
@@ -85,60 +85,443 @@ NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
+* __permissions__: Array of granted permissions (read, write, share)
+* __protocol__: Access protocol configuration (webdav, webapp, datatx)
+* __providerId__: Unique identifier for the Share at the provider
+* __requirements__: Array of access requirements (must-use-mfa,
```
Why would I need to store those with the (Sent)Share?

mickenordin replied:
These are part of a share object according to the spec, and at least as sending server you need to for example keep track of what permissions was mandated for the share, no?

KrausMatthias commented on IETF-RFC.md:
```diff
@@ -85,60 +85,443 @@ NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
+* __expiration__: Optional expiration timestamp
+* __name__: Human-readable name of the shared Resource
+* __owner__: OCM Address of the Resource owner
+* __permissions__: Array of granted permissions (read, write, share)
```
In my eyes permissions currently are protocol specific (at least in the Share Creation Notification) and is stored in the protocol configuration?

KrausMatthias commented on IETF-RFC.md:
```diff
@@ -166,48 +549,63 @@ characters.
+  - acts as an API client, allowing the receiving user to access the
+    Resource through an API (e.g., WebDAV [RFC4918]) of the sending
+    server.
+* __Remote Resource__ - A Resource provided by the Sending Server.
```
A Resource accessible on a Receiving Server via a Share from a Sending Server ?

KrausMatthias commented on IETF-RFC.md:
```diff
@@ -166,48 +549,63 @@ characters.
   Receiving Party on the modes of accessing the Resource.
 * __Share Requirements__ - Protocol-specific restrictions on the modes
   of accessing the Resource.
+* __Shared Resource__ - A Resource shared by an OCM Server, becoming a
+  Remote Resource if accepted by the Invite Receiver OCM Server.
```
Whats the difference to a Resource?

mickenordin replied:
These two (this and your comment above) have been part of the definitions for a long time, I just alphabetized them. But I take your point that there is a lot of redundancy in definitions, and I was also pondering that when sorting them.

KrausMatthias commented on IETF-RFC.md:
````diff
@@ -315,6 +703,34 @@ Share Creation Notification from an unknown Sending Party is received
 * the Invite Receiver OCM Server sends the Invite Acceptance Request to
   the Invite Sender OCM Server
 
+```
````
Add Users and their actions (Alice and Bob :D )?

KrausMatthias reviewed (changes requested):
This is a great improvement! Adds a lot of clarity to the involved concepts.

I think a distinction on what is Binding Specification and what is helpful context should be added.

mickenordin committed:
Address feedback from @KrausMatthias

KrausMatthias commented:
I think my main concern is still what is strictly necessary for interoperability, and what should be left open to implementations (while guidance regarding the concepts is still extremely helpful there). 

So I'm not after removing details but clearly separating, "this must what a share payload looks like with these semantics so it works with other implementations" and "this is what you might keep track of in your implementation to build a working service, but doing it differently would still work". 

So maybe we could have two separate sections? "Protocol Specification" and "Context and Implementation Guidance" ?

glpatcern commented:
> I think my main concern is still what is strictly necessary for interoperability, and what should be left open to implementations (while guidance regarding the concepts is still extremely helpful there).

Overall I'm quite sympathetic with this comment. I appreciate the overall effort to "canonicalize" the implementations, but it has to be clear what is prescribed and what is guidance. I did not have time yet to review this PR in details and I thank both of you for the combined effort!

glpatcern commented on IETF-RFC.md:
````diff
@@ -315,6 +683,34 @@ Share Creation Notification from an unknown Sending Party is received
+                                                 |   Added     |
+                                                 +-------------+
+```
+
````
Here I'd either include the mermaid-formatted Sequence Diagrams we have in the repo or refer to them, as opposed to try and draw them.

glpatcern commented on IETF-RFC.md:
```diff
@@ -166,48 +529,63 @@ characters.
+  Shares; similar to "Resource Owner" in OAuth [RFC6749], identified by
+  its OCM Address.
+* __Sending Server__ - The server that:
+  - holds the Resource ("file server" or "Entreprise File Sync and Share
```
I'd remove the `file server or EFSS` example, or explicitly state `for instance, a file server`.

glpatcern commented on IETF-RFC.md:
```diff
@@ -85,60 +85,423 @@ NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
+Resources. It MAY provide it's users with an _Address Book_ of
+_Contacts_ and the ability to accept _Invites_.
+
+## Object models
```
I'm not so convinced that this section should be at the very beginning of the Draft. What do other people think?

mickenordin replied:
We can put it in an appendix, maybe that is better?

glpatcern replied:
Yes, that seems better indeed

glpatcern reviewed (commented):
I started to give a full review and already have a few comments, possibly more to come, but overall I think it's an effort in the good direction

mickenordin committed:
Remove sequence diagram better supplied elsewhere

mickenordin committed:
Move object model to appendix D

mickenordin commented:
> > I think my main concern is still what is strictly necessary for interoperability, and what should be left open to implementations (while guidance regarding the concepts is still extremely helpful there).
> 
> Overall I'm quite sympathetic with this comment. I appreciate the overall effort to "canonicalize" the implementations, but it has to be clear what is prescribed and what is guidance.

I think many of these concerns are addressed now, by moving it to an appendix and stating that it is non-normative.

glpatcern commented on IETF-RFC.md:
```diff
@@ -85,60 +85,96 @@ NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
-  Resource, after establishing contact with an OCM Server.
-* __Discovering Server__ - A server that tries to obtain information in
-  OCM API Discovery.
+Open Cloud Mesh defines distinct functions, it is not neccessary for an
```
```suggestion
Open Cloud Mesh defines distinct functions.  It is not necessary for an
```

glpatcern commented on IETF-RFC.md:
```diff
@@ -85,60 +85,96 @@ NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
+
+A Sending Server is an OCM Provider that holds Resources and exposes
+APIs to allow access to them.  It allows its users to create _Shares_
+to give other users access to those Resources. A Sending Server MAY
```
```suggestion
to give other users access to those Resources.  A Sending Server MAY
```

glpatcern commented on IETF-RFC.md:
```diff
@@ -85,60 +85,96 @@ NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
+APIs to allow access to them.  It allows its users to create _Shares_
+to give other users access to those Resources. A Sending Server MAY
+provide its users with the ability to generate _Invites_ to establish
+contact with other users on other OCM Providers. When doing so it MAY
```
```suggestion
contact with other users on other OCM Providers.  When doing so it MAY
```

glpatcern commented on IETF-RFC.md:
```diff
@@ -85,60 +85,96 @@ NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED",
+A Receiving Server is an OCM Provider that receives _Share_ Creation
+Notifications from Sending Servers, notifies its users about incoming
+_Shares_, and acts as an API client to allow its users to access Remote
+Resources. It MAY provide it's users with an _Address Book_ of
```
```suggestion
Resources.  It MAY provide its users with an _Address Book_ of
```

glpatcern commented on IETF-RFC.md:
```diff
@@ -218,8 +268,9 @@ characters.
-  trusted OCM Servers.
+
+In Appendix D, an object model is presented as a non-normative guide for
+implementators to understand the relationships between these terms.
```
```suggestion
implementers to understand the relationships between these terms.
```

glpatcern commented on IETF-RFC.md:
```diff
@@ -1394,6 +1438,337 @@ format:
+
+An implementor of OCM MAY choose any internal object model to represent
+an _Address Book_, a _Contact_, an _Invite_, a _Provider_, a _Share_,
+and  a _User_.  However, the following diagrams are provided to clarify
```
```suggestion
and a _User_.  The following diagrams are provided to clarify
```

glpatcern commented on IETF-RFC.md:
```diff
@@ -1394,6 +1438,337 @@ format:
+Party_ in preparation for Share Creation.
+
+The Address Book entity maintains a collection of contacts for a user
+within the OCM provider. It serves as the primary mechanism for managing
```
Here and in the following we should still follow the IETF convention of two-spaces-after-full-stop

glpatcern reviewed (changes requested):
I think this looks good now, the requested changes are just spelling mistakes.

mickenordin committed:
Address feedback from Giuseppe

Co-authored-by: Giuseppe Lo Presti <giuseppe.lopresti@cern.ch>

glpatcern reviewed (dismissed):
I'm happy with this now. @KrausMatthias what about you?

KrausMatthias reviewed (commented):
Should we distinguish between Sent Shares and Received Shares in the Object models?

Otherwise I think it's perfect :)

MahdiBaghbani reviewed (dismissed):
I've read it now, It's great, nice ascii flowcharts! Thanks Micke

glpatcern commented:
> Should we distinguish between Sent Shares and Received Shares in the Object models?

True that in a real implementation (e.g. in Reva) those two types are expected to be quite different, at least because in the Received Share the remote URI has to be stored. We may want to emphasize that.

mickenordin commented:
> Should we distinguish between Sent Shares and Received Shares in the Object models?
> 
> Otherwise I think it's perfect :)

This is a great point, I will add a 'direction' to the model.

mickenordin commented:
I added a direction to the share model now

glpatcern commented on IETF-RFC.md:
```diff
@@ -1419,32 +1416,361 @@ adhere to the following format:
+* __expiration__: Optional expiration timestamp
+* __name__: Human-readable name of the shared Resource
+* __owner__: OCM Address of the Resource owner
+* __protocol__: Access protocol configuration (webdav, webapp, datatx)
```
```suggestion
* __protocol__: Access protocol name and details (webdav, ssh, webapp, ...)
```

glpatcern commented on IETF-RFC.md:
```diff
@@ -1419,32 +1416,361 @@ adhere to the following format:
++------------------+
+|   Protocols      |
++------------------+
+| - datatx         |
```
```suggestion
| - ssh            |
```

glpatcern commented on IETF-RFC.md:
```diff
@@ -1419,32 +1416,361 @@ adhere to the following format:
+            |    (OCM Server)       |
+            +-----------------------+
+            | - apiVersion          |
+            | - enabled: bool       |
```
```suggestion
            | - enabled             |
```

glpatcern reviewed (changes requested):
The changes below are trivial fixes + the removal of `datatx`.
Otherwise, concerning the `direction`, I'm not so sure it helps at this level of abstraction. As a matter of fact, the `protocols` encapsulate the access details, including whether the share is incoming or sent, and a real model would have other more significant differences (cf. for example in Reva the model for a [sent share](https://github.com/cs3org/reva/blob/master/pkg/share/manager/sql/model/model.go#L161-L175) vs an [incoming share](https://github.com/cs3org/reva/blob/master/pkg/share/manager/sql/model/model.go#L186-L201)).
As such, I'd vote for removing the `direction` (stating that it's part of the access protocols), or for further detailing the model as secondary option.

mickenordin commented:
> The changes below are trivial fixes + the removal of `datatx`. Otherwise, concerning the `direction`, I'm not so sure it helps at this level of abstraction. As a matter of fact, the `protocols` encapsulate the access details, including whether the share is incoming or sent, and a real model would have other more significant differences (cf. for example in Reva the model for a [sent share](https://github.com/cs3org/reva/blob/master/pkg/share/manager/sql/model/model.go#L161-L175) vs an [incoming share](https://github.com/cs3org/reva/blob/master/pkg/share/manager/sql/model/model.go#L186-L201)). As such, I'd vote for removing the `direction` (stating that it's part of the access protocols), or for further detailing the model as secondary option.

Fixed the nits, and removed direction

mickenordin committed:
Merge branch 'develop' into kano-definitions

mickenordin commented:
I had to fix a merge confict, and that made the reviews stale... Can you please re-review?

glpatcern reviewed (commented):
Yep, I got there was a conflict, but are you sure this is now the final state? It seems the paragraph about the `Functions` is gone

mickenordin committed:
Add back functions and roles

mickenordin commented:
> Yep, I got there was a conflict, but are you sure this is now the final state? It seems the paragraph about the `Functions` is gone

I added back functions and roles, does it look ok now?

glpatcern reviewed (approved):
Yes!
